### PR TITLE
[ADD] Documentation update in `base_trainer.py`

### LINKED ISSUE #406: Documentation update
https://github.com/automl/Auto-PyTorch/blob/0cc6b4696abcd2e912ebd4fc9a6c283ef5607b2e/autoPyTorch/pipeline/components/training/trainer/base_trainer.py#L229

The class documentation should be updated. As I checked, the module also is in need of a documentation update.

### PR DESCRIPTION
**Pull request by theodorju**
## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

## Description

I have updated the documentation of `base_trainer` for #406 in places where it seemed reasonable to add documentation.

## Motivation and Context

Changes fix #406 
## How has this been tested?

Since the changes are only in the documentation, I have not executed any tests for this issue.

### CODE REVIEW

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 1]
```diff
@@ -44,14 +44,28 @@ def __init__(self,
...
     def is_max_epoch_reached(self, epoch: int) -> bool:
+        """
+        For budget type 'epoch' or 'epoch_or_time' return True if the maximum
+            number of epochs is reached.
+        Args:
+            epoch (int): the current epoch
 
+        Returns:
+            True if the current epoch is larger than the maximum epochs, False otherwise.
+            Additionally, returns False if the run is without this constrain.
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:67]:
```suggestion
            Additionally, returns False if the run is without this constraint.
```

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 2]
```diff
@@ -60,6 +74,13 @@ def is_max_epoch_reached(self, epoch: int) -> bool:
         return False
 
     def is_max_time_reached(self) -> bool:
+        """
+        For budget type 'runtime' or 'epoch_or_time' return True if the maximum
+            runtime is reached.
+        Returns:
+            True if the maximum runtime is reached, False otherwise.
+            Additionally, returns False if the run is without this constrain.
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:82]:
```suggestion
            Additionally, returns False if the run is without this constraint.
```

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 3]
```diff
@@ -44,14 +44,28 @@ def __init__(self,
 
         It also allows to define a 'epoch_or_time' budget type, which means,
         the first of them both which is exhausted, is honored
+        Args:
+            budget_type (str): Type of budget to be used when fitting the pipeline.
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:48]: Could you add a new line and indentation to the description of `budget_type`? You can check the base_task.py file for how we format the docstrings.

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 4]
```diff
@@ -44,14 +44,28 @@ def __init__(self,
...
+                Possible values are 'epochs', 'runtime', or 'epoch_or_time'
+            max_epochs (Optional[int], default=None): Maximum number of epochs to train the pipeline for
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:50]: Could you also do this for all the other parameter descriptions?

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 5]
```diff
@@ -118,8 +144,20 @@ def add_performance(self,
                         test_loss: Optional[float] = None,
                         ) -> None:
         """
-        Tracks performance information about the run, useful for
-        plotting individual runs
+        Tracks performance information about the run, useful for plotting individual runs.
+        Args:
+            epoch (int): the current epoch
+            start_time (float): timestamp at the beginning of current epoch
+            end_time (float): timestamp when gathering the information
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:151]:
```suggestion
            end_time (float): timestamp when gathering the information after the current epoch.
```

**Review by ravinkohli — CHANGES_REQUESTED** [review]
Hey, thanks for the pull request. I have left some minor comments after which I think we can merge this PR.

**Comment by theodorju** [comment]
Thanks for the comments. I have applied all the suggested changes. Additionally:
- I have kept the 120 max line width, and also modified some pre-existing docstrings accordingly
- Added more detailed descriptions in return sections, where applicable
- Updated some docstrings that I've missed in the previous commit (e.g. [`criterion_preparation`](https://github.com/theodorju/Auto-PyTorch/blob/953b3a086b81d40a7fc26a9621f521df4732eb78/autoPyTorch/pipeline/components/training/trainer/base_trainer.py#L708))

**Pull request closed without merging**